src/main/scala/cromwell/engine/workflow/WorkflowManagerActor.scala

```diff
@@ -62,32 +71,23 @@ object WorkflowManagerActor {
  * WorkflowOutputs: Returns a `Future[Option[binding.WorkflowOutputs]]` aka `Future[Option[Map[String, WdlValue]]`
  *
  */
-class WorkflowManagerActor(backend: Backend) extends Actor with CromwellActor {
+class WorkflowManagerActor(backend: Backend) extends LoggingFSM[WorkflowManagerState, WorkflowManagerData] with CromwellActor {
   import WorkflowManagerActor._
-  private val log = Logging(context.system, this)
+  private val logger = Logging(context.system, this)
   private val tag = "WorkflowManagerActor"
 
   type WorkflowActorRef = ActorRef
 
   private val workflowStore = new WriteOnceStore[WorkflowId, WorkflowActorRef]
 
-  if (getAbortJobsOnTerminate) {
-    Runtime.getRuntime.addShutdownHook(new Thread() {
-      override def run(): Unit = {
-        log.info(s"$tag: Received shutdown signal. Aborting all running workflows...")
-        workflowStore.toMap.foreach{case (id, actor)=>
-          CromwellServer.workflowManagerActor ! WorkflowManagerActor.WorkflowAbort(id)
-        }
-        var numRemaining = -1
-        while(numRemaining != 0) {
-          Thread.sleep(1000)
-          val result = globalDataAccess.getWorkflowsByState(Seq(WorkflowRunning, WorkflowAborting))
-          numRemaining = Await.result(result,Duration.Inf).size
-          log.info(s"$tag: Waiting for all workflows to abort ($numRemaining remaining).")
-        }
-        log.info(s"$tag: All workflows aborted.")
-      }
-    })
+  private val donePromise = Promise[Unit]()
+
+  if (abortJobsOnTerminate) {
+    sys.addShutdownHook {
```

Comment:
Ah gotcha.  I hadn't considered that, I think you're right we may be able to get away with only one hook in WMA.  SWRA's `WorkflowAborted` handler would need to be modified to allow for that, but I think that should be done anyway as the current SWRA abort handler looks wrong (it notes that the only workflow it cares about has been aborted and then tells no other actors and hangs out forever doing nothing).